crypto/ascon/ascon_aead128.c

```diff
@@ -0,0 +1,436 @@
+            ASCONP12(ctx->state[0], ctx->state[1], ctx->state[2], ctx->state[3],
+                     ctx->state[4]);
+            ctx->offset = 0;
+        }
+        /* sponge: absorb a message byte */
+        ctx->state[0] ^= (uint64_t)(*m++) << 8 * ctx->offset++;
+    }
+}
+
+void ascon_hash256_final(ascon_hash256_ctx *ctx, unsigned char *digest)
+{
+    uint64_t s0, s1, s2, s3, s4;
+    unsigned char pad = 0x01;
+
+    /* message termination */
+    ascon_hash256_update(ctx, &pad, 1);
+
+    s0 = ctx->state[0];
+    s1 = ctx->state[1];
+    s2 = ctx->state[2];
+    s3 = ctx->state[3];
+    s4 = ctx->state[4];
+
+    /* sponge: squeeze out four words for the hash */
+    ASCONP12(s0, s1, s2, s3, s4);
+    PUTU64(digest, s0);
+    ASCONP12(s0, s1, s2, s3, s4);
+    PUTU64(digest + 8, s0);
+    ASCONP12(s0, s1, s2, s3, s4);
+    PUTU64(digest + 16, s0);
+    ASCONP12(s0, s1, s2, s3, s4);
+    PUTU64(digest + 24, s0);
+}
+
+#ifdef OPENSSL_BUILDING_OPENSSL
+/* Provider compatibility wrapper functions */
+void ossl_ascon_aead128_init(ascon_aead_ctx_t *ctx, const unsigned char *k,
+                             const unsigned char *n)
+{
+    ascon_aead128_init(ctx, k, n);
+}
+
+void ossl_ascon_aead128_assoc_data_update(ascon_aead_ctx_t *ctx,
+                                          const unsigned char *in, size_t inl)
+{
+    ascon_aead128_aad_update(ctx, in, inl);
+}
+
+size_t ossl_ascon_aead128_encrypt_update(ascon_aead_ctx_t *ctx,
+                                         unsigned char *out,
+                                         const unsigned char *in, size_t inl)
+{
+    ascon_aead128_encrypt_update(ctx, out, in, inl);
+    return inl;
+}
+
+size_t ossl_ascon_aead128_decrypt_update(ascon_aead_ctx_t *ctx,
+                                         unsigned char *out,
+                                         const unsigned char *in, size_t inl)
+{
+    ascon_aead128_decrypt_update(ctx, out, in, inl);
+    return inl;
+}
+
+size_t ossl_ascon_aead128_encrypt_final(ascon_aead_ctx_t *ctx,
+                                        unsigned char *out,
+                                        unsigned char *tag, size_t tag_len)
+{
+    unsigned char computed_tag[16];
+
+    ascon_aead128_final(ctx, computed_tag);
+    if (tag != NULL && tag_len >= 16)
+        memcpy(tag, computed_tag, 16);
+    if (out != NULL) {
+        /* No additional output for final */
+    }
+    return 0;
+}
+
+size_t ossl_ascon_aead128_decrypt_final(ascon_aead_ctx_t *ctx,
+                                        unsigned char *out,
+                                        bool *is_tag_valid,
+                                        const unsigned char *tag,
+                                        size_t tag_len)
+{
+    unsigned char computed_tag[16];
+
+    ascon_aead128_final(ctx, computed_tag);
+    if (is_tag_valid != NULL && tag != NULL && tag_len >= 16) {
+        *is_tag_valid = (memcmp(computed_tag, tag, 16) == 0);
```

Comment:
Shouldn't this be done in constant time? Otherwise we allow an attacker to potentially forge a tag, turning an AEAD into a malleable cypher